lib/cocoapods-core/cdn_source.rb

```diff
@@ -369,104 +386,112 @@ def download_and_save_with_retries_async(partial_url, file_remote_url, etag, ret
-          # optimization. See #initialize for more information.
-          FileUtils.touch path
-          partial_url
-        when 200
-          File.open(path, 'w') { |f| f.write(response.response_body.force_encoding('UTF-8')) }
-
-          etag_new = response.headers['etag'] unless response.headers.nil?
-          debug "CDN: #{name} Relative path downloaded: #{partial_url}, save ETag: #{etag_new}"
-          File.open(etag_path, 'w') { |f| f.write(etag_new) } unless etag_new.nil?
-          partial_url
-        when 404
-          debug "CDN: #{name} Relative path couldn't be downloaded: #{partial_url} Response: #{response.response_code}"
-          nil
-        when 502, 503, 504
-          # Retryable HTTP errors, usually related to server overloading
-          if retries <= 1
-            raise Informative, "CDN: #{name} URL couldn't be downloaded: #{file_remote_url} Response: #{response.response_code} #{response.response_body}"
-          else
-            debug "CDN: #{name} URL couldn't be downloaded: #{file_remote_url} Response: #{response.response_code} #{response.response_body}, retries: #{retries - 1}"
-            exponential_backoff_async(retries).then do
-              download_and_save_with_retries_async(partial_url, file_remote_url, etag, retries - 1)
-            end
+      response =
+        begin
+          Async::Task.current.with_timeout(REQUEST_TIMEOUT) do
+            create_download_task(file_remote_url, etag)
           end
-        when 0
-          # Non-HTTP errors, usually network layer
+        rescue Async::TimeoutError, Async::HTTP::Protocol::RequestFailed, SocketError, ::StandardError => e
+          message =
+            case e
+            when Async::TimeoutError
+              "CDN: #{name} URL couldn't be downloaded: #{file_remote_url} Response: Request timeout"
+            when SocketError
+              "CDN: #{name} URL couldn't be downloaded: #{file_remote_url} Response: Couldn't connect to server"
+            else
+              "CDN: #{name} URL couldn't be downloaded: #{file_remote_url} Response: #{e.message}"
+            end
+
           if retries <= 1
-            raise Informative, "CDN: #{name} URL couldn't be downloaded: #{file_remote_url} Response: #{response.return_message}"
+            raise Informative, message
           else
-            debug "CDN: #{name} URL couldn't be downloaded: #{file_remote_url} Response: #{response.return_message}, retries: #{retries - 1}"
-            exponential_backoff_async(retries).then do
-              download_and_save_with_retries_async(partial_url, file_remote_url, etag, retries - 1)
+            debug message + ", retries: #{retries - 1}"
+            sleep_async backoff_time(retries)
+            download_and_save_with_retries_async(partial_url, file_remote_url, etag, retries - 1)
+            return
+          end
+        end
+
+      body = response.read
+      case response.status
+      when 301
```

Comment:
@jasl check out https://github.com/CocoaPods/Core/pull/679 - could you please add `302` here?

Reply:
Added, I also rebased master and squashed commits